### ISSUE TITLE
Override path for .terraform.lock.hcl

### ISSUE THREAD

**Issue by ademariag** [hashicorp/terraform#27241]
### Current Terraform Version
```
terraform --version
Terraform v0.14.2
```

### Use-cases
I am one of the founders of the https://kapitan.dev project and we use Kapitan to generate (json) terraform config files using jsonnet.

The recent introduction of the `.terraform.lock.hcl` breaks our workflows (and any other terraform user of Kapitan) because  the file is rewritten every time we compile a target.

### Proposal

If you could provide a way to store the file under another path (still in git as you suggest, but a path where it is not deleted) that would be awesome.

For example, currently terraform files could be under:
```
compiled/
  production/
    terraform/
      .terraform.lock.hcl
      ...other hcl files...
```

Ideally we would like to pass the correct path using an env variable, i.e. TERRAFORM_LOCK_FILE_PATH
```
export TERRAFORM_LOCK_FILE_PATH=$(git rev-parse --show-toplevel)/terraform/$TARGET_NAME/
```

 We could then store the files somewhere else:
```
terraform/   <---- not destroyed on compile
  production/
      .terraform.lock.hcl
compiled/
  production/
    terraform/
      ...other hcl files...
```

**Comment by stazz** [comment]
I have similar need for this, but for different reason:

Within the infra pipelines we are running Terraform from, we are exclusively doing it using Docker (among many things, it allows for easy tool dependency management and security model in our pipelines). It is extremely important for us to verify that pipelines do not introduce any extra changes to the repository extracted at certain commit by the pipeline. This is why we pass our TF code as **readonly** Docker volume mount to the TF container - certainly running TF shouldn't do any changes to our TF code, right?

Starting with 0.14 version and introduction of this dependency lock file, the `terraform init` command fails to write the dependency lock file:

```
Error while writing new dependency lock information to .terraform.lock.hcl:
cannot create temporary file to update .terraform.lock.hcl: open .terraform.lock.hcl217706229: read-only file system.
```

While dependency lock file as an idea is awesome, I agree with @ademariag that it should be a way to customize the actual location of the file within the file system.

**Comment by bigpapabear805** [comment]
Can someone please do something it's starting to hurt my lower back

**Comment by starkers** [comment]
Yes please, the clash with the lock file has been extremely painful for my team and we've had to look to revert terraform

**Comment by apparentlymart** [comment]
It seems like there are at least two different situations represented here:

* When using Terraform with code generation tools rather than hand-written configuration, it's better to keep the lock file with the input files rather than the generated files.
* When the working directory is on a read-only filesystem, it's inconvenient that Terraform wants to rewrite the lock file even if the new content would be identical anyway, because there are no new provider dependencies.

I'm considering these separately because it seems like the solutions for them are likely to be separate too. For the second, it would be best if no special Terraform settings were required to run in a read-only directory, so I'd prefer not to address that by forcing to relocate the file from the default location.

**Comment by thusoy** [comment]
I'm having the same issue with terraform running on a read-only filesystem.

Couple options that would work for me at least:
1) Envvar or cli flag to disable updating the lockfile
2) Gracefully ignore updating the lockfile on permission denied (or write errors in general)
3) Specify path to lockfile. I'm actually being hit by both this issue and https://github.com/hashicorp/terraform/issues/27158 (we have the same setup with a directory of symlinks for each environment), if I could specify the path I wouldn't need the lockfile symlink).

**Comment by toadjaune** [comment]
I have a third use-case : 

I have a project with 3 different environments (test, preprod, prod), and each of them is in a separate directory/tfstate. Most of the comfig is shared via a module.

I'd like to have single lockfile for all three environments, in order to guarantee that the config validated on test environment is applied similarly on production.

(I think workspaces could allow us to work around the issue in this specific usecase, but due to various constraints, we chose not to go down this path when implementing this project)

This usecase is actually pretty close from the one when terrafom is used with generation tools.
Any of the mentioned solutions (cli argument, envvar, configuration directive in manifests) would work here (with a preference for the latter, as we wouldn't want a user to forget specifying the lockfile path)

**Comment by lorengordon** [comment]
> 1. Envvar or cli flag to disable updating the lockfile

I'm using `terraform-bundle`, and would actually just like to disable the lock-file generation entirely (not just updating it).

**Comment by awendt** [comment]
@apparentlymart 

> When the working directory is on a read-only filesystem, it's inconvenient that Terraform wants to rewrite the lock file even if the new content would be identical anyway, because there are no new provider dependencies.

To be more precise, this is more than an inconvenience to me – I find it confusing and unexpected that Terraform attempts to write the lock file on anything other than `terraform init -upgrade`. I understand the file is Terraform's business in terms of content, but I would expect the point in time when the file is written to be under my control.

**Comment by stazz** [comment]
Can confirm this still happens with `0.14.7` and `0.15.0-beta1` versions. Any progress on this issue, @apparentlymart ? :)

**Comment by apparentlymart** [comment]
In Terraform's default configuration, `terraform init` will generate a lock file containing all of the checksums signed by the original provider author across all platforms. However, it's true that currently if you override that behavior to make Terraform install from somewhere other than the origin registry (either via explicit installation method configuration or by enabling the read-through cache) then `terraform init` becomes constrained by the limited metadata available on those alternate installation methods, in which case you need to use `terraform providers lock` to achieve a result like `terraform init` would produce by default.

Let's keep this issue focused on the problem of where Terraform writes the file and when Terraform tries to update the file, because there are other issues open about what Terraform will record in the file under different CLI configurations and so we won't be addressing those behaviors in this issue.

**Comment by dee-kryvenko** [comment]
@apparentlymart that is not true - please check it out in https://github.com/hashicorp/terraform/issues/28041#issuecomment-881083013. Neither `tf init` nor `tf providers lock` are producing cross-platform lock. You are probably confusing different hash types i.e. `zh` and `h1` thinking these are cross-platform entries. Or maybe cross-platform entries works only - I am not sure. Either way for a simple

```hcl
resource "null_resource" "test" {}
```

`tf init` and `tf provider lock` both generates:

```hcl
# This file is maintained automatically by "terraform init".
# Manual edits may be lost in future updates.

provider "registry.terraform.io/hashicorp/null" {
  version = "3.1.0"
  hashes = [
    "h1:xhbHC6in3nQryvTQBWKxebi3inG5OCgHgc4fRxL0ymc=",
    "zh:02a1675fd8de126a00460942aaae242e65ca3380b5bb192e8773ef3da9073fd2",
    "zh:53e30545ff8926a8e30ad30648991ca8b93b6fa496272cd23b26763c8ee84515",
    "zh:5f9200bf708913621d0f6514179d89700e9aa3097c77dac730e8ba6e5901d521",
    "zh:9ebf4d9704faba06b3ec7242c773c0fbfe12d62db7d00356d4f55385fc69bfb2",
    "zh:a6576c81adc70326e4e1c999c04ad9ca37113a6e925aefab4765e5a5198efa7e",
    "zh:a8a42d13346347aff6c63a37cda9b2c6aa5cc384a55b2fe6d6adfa390e609c53",
    "zh:c797744d08a5307d50210e0454f91ca4d1c7621c68740441cf4579390452321d",
    "zh:cecb6a304046df34c11229f20a80b24b1603960b794d68361a67c5efe58e62b8",
    "zh:e1371aa1e502000d9974cfaff5be4cfa02f47b17400005a16f14d2ef30dc2a70",
    "zh:fc39cc1fe71234a0b0369d5c5c7f876c71b956d23d7d6f518289737a001ba69b",
    "zh:fea4227271ebf7d9e2b61b89ce2328c7262acd9fd190e1fd6d15a591abfa848e",
  ]
}
```

Note there's single `h1` entry, and it is for darwin. If I run it on linux - tf will add a second `h1` entry to it. And it will fail if it was using local cache, which is a separate ticket https://github.com/hashicorp/terraform/issues/28041.

**Comment by marianod92** [comment]
> The only temporary solution that we've found is to make it a symlink and replace target file
> 
> The only huge issue is that terraform **replaces the file**. Literally, instead of changing content it replaces symlink with a file, so we need to move and change it back to symlink

Hi @nick4fake 

But do you do this action manually or with a wrapper?

For example, [Terragrunt](https://terragrunt.gruntwork.io/docs/features/lock-file-handling/#how-terragrunt-solves-this-problem) mentions this problem too and just moves the path file.

Thanks in advance!

**Comment by apparentlymart** [comment]
Hi all,

I just wanted to update to say that Terraform v0.15.0 and later have an extra option `-lockfile=readonly` for `terraform init` which will tell Terraform to use the existing lock file without modification if possible, and to fail with an error if the existing lock file is insufficient to describe dependencies for the current configuration. If you combine that with the existing `TF_DATA_DIR` environment variable it should be possible to run `terraform init` on a read-only filesystem as long as the lock file is kept up-to-date in your version control system by developers.

**Comment by ademariag** [comment]
hello @apparentlymart ,

can you explain better how to combine the `TF_DATA_DIR` with this suggestion? I read from https://developer.hashicorp.com/terraform/language/files/dependency-lock#lock-file-location that the expected location is still the current working directory
